src/EditorFeatures/Core/Tagging/TaggerThreadCoordinator.cs

```diff
@@ -0,0 +1,213 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Tagging;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Tagging
+{
+    /// <summary>
+    /// Helper type that is used by all <see cref="AsynchronousTaggerProvider{TTag}"/>s to coordinator sending messages
+    /// to the UI thread to do work.  The main problem this wants to avoid is N taggers waking up at different times,
+    /// sending UI thread messages both to collect UI information, and then UI messages to notify about updated tags.
+    /// By utilizing a shared <see cref="TaggerThreadCoordinator"/> all the taggers can instead enqueue that work into 
+    /// a shared queue, which then executes as many of those UI items in the same UI message as long as they fit within
+    /// the same timeslice.
+    /// </summary>
+    [Export(typeof(TaggerThreadCoordinator)), Shared]
+    internal sealed class TaggerThreadCoordinator
```

Comment:
Out of curiosity, does this need to be tied to tagging, or could this be used more generally?

Reply:
it coudl be used more generally.  However, tagging is the area where we have N features taht often wake up, do work (using data obtained from UI thread), then want to report changes on the UI thread.

Reply:
Just seemed like a generally useful concept that could be nice to use outside the tagging area. I guess we could move it later if we find other pieces of code that are frequently trying to get to the UI thread.